### DISCUSSION TITLE
verify_convergence option in optimize_yaw_angles not working

### DISCUSSION THREAD

**Discussion by fwasilcz** [NatLabRockies/floris#1015] [Q&A]
Hi,
I run into some problems during yaw optimization excercise and to get to the bottom of it I tried turning on "verify_convergence" option in optimize_yaw_angles method. 
My first question is - how is the convergence verified, I did not find anything about it.
Secondly, when I turned the option on, the optimizer returns an error (posted below). The error is generated only by the "verify convergence" option, with it turned off the code works fine, so I'm guessing this may be a bug.

Optimizing yaw angles with 16 workers.
_RemoteTraceback: 
"""
Traceback (most recent call last):
  File "C:\Users\Filip_Win\miniconda3\envs\floris4.0\Lib\concurrent\futures\process.py", line 263, in _process_worker
    r = call_item.fn(*call_item.args, **call_item.kwargs)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Filip_Win\miniconda3\envs\floris4.0\Lib\concurrent\futures\process.py", line 212, in _process_chunk
    return [fn(*args) for args in chunk]
            ^^^^^^^^^
  File "C:\Users\Filip_Win\miniconda3\envs\floris4.0\Lib\site-packages\floris\parallel_floris_model.py", line 48, in _optimize_yaw_angles_serial
    df_opt = yaw_opt.optimize(print_progress=print_progress)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Filip_Win\miniconda3\envs\floris4.0\Lib\site-packages\floris\optimization\yaw_optimization\yaw_optimizer_sr.py", line 310, in optimize
    df_opt = self._finalize()
             ^^^^^^^^^^^^^^^^
  File "C:\Users\Filip_Win\miniconda3\envs\floris4.0\Lib\site-packages\floris\optimization\yaw_optimization\yaw_optimization_base.py", line 388, in _finalize
    self._verify_solutions_for_convergence(
  File "C:\Users\Filip_Win\miniconda3\envs\floris4.0\Lib\site-packages\floris\optimization\yaw_optimization\yaw_optimization_base.py", line 566, in _verify_solutions_for_convergence
    jj = (ids_max_loss[0][0], ids_max_loss[1][0])
                              ~~~~~~~~~~~~^^^
IndexError: tuple index out of range
"""


The above exception was the direct cause of the following exception:

Traceback (most recent call last):

  File ~\miniconda3\envs\floris4.0\Lib\site-packages\spyder_kernels\py3compat.py:356 in compat_exec
    exec(code, globals, locals)

  File d:\wyniki\wiatr\analizy_ekonomiczne\definicja_farmy\floris4.0\calculate_and_save_optimal_yaw_10deg_simple.py:118
    df_opt = pfmodel.optimize_yaw_angles(

  File ~\miniconda3\envs\floris4.0\Lib\site-packages\floris\parallel_floris_model.py:509 in optimize_yaw_angles
    df_opt = pd.concat(df_opt_splits, axis=0).reset_index(drop=True).sort_values(

  File ~\miniconda3\envs\floris4.0\Lib\site-packages\pandas\core\reshape\concat.py:382 in concat
    op = _Concatenator(

  File ~\miniconda3\envs\floris4.0\Lib\site-packages\pandas\core\reshape\concat.py:445 in __init__
    objs, keys = self._clean_keys_and_objs(objs, keys)

  File ~\miniconda3\envs\floris4.0\Lib\site-packages\pandas\core\reshape\concat.py:504 in _clean_keys_and_objs
    objs_list = list(objs)

  File ~\miniconda3\envs\floris4.0\Lib\concurrent\futures\process.py:642 in _chain_from_iterable_of_lists
    for element in iterable:

  File ~\miniconda3\envs\floris4.0\Lib\concurrent\futures\_base.py:619 in result_iterator
    yield _result_or_cancel(fs.pop())

  File ~\miniconda3\envs\floris4.0\Lib\concurrent\futures\_base.py:317 in _result_or_cancel
    return fut.result(timeout)

  File ~\miniconda3\envs\floris4.0\Lib\concurrent\futures\_base.py:449 in result
    return self.__get_result()

  File ~\miniconda3\envs\floris4.0\Lib\concurrent\futures\_base.py:401 in __get_result
    raise self._exception

IndexError: tuple index out of range


Verifying convergence of the found optimal yaw angles.
Verifying convergence of the found optimal yaw angles.
Verifying convergence of the found optimal yaw angles.
Verifying convergence of the found optimal yaw angles.
Verifying convergence of the found optimal yaw angles.
Verifying convergence of the found optimal yaw angles.
Verifying convergence of the found optimal yaw angles.
Verifying convergence of the found optimal yaw angles.
Verifying convergence of the found optimal yaw angles.
Verifying convergence of the found optimal yaw angles.
Verifying convergence of the found optimal yaw angles.
Verifying convergence of the found optimal yaw angles.
Verifying convergence of the found optimal yaw angles.
Verifying convergence of the found optimal yaw angles.
Verifying convergence of the found optimal yaw angles.
Verifying convergence of the found optimal yaw angles.

**Comment by misi9170** [comment]
Hi @fwasilcz , 

Thank you for your question. If I'm understanding correctly, you are using the `optimize_yaw_angles()` method on the `ParallelFlorisModel` (with `verify_convergence=True`) when you run into this issue. This does look like a bug: `verify_convergence` is simply passed on to the [`YawOptimizationSR` class](https://github.com/NREL/floris/blob/main/floris/optimization/yaw_optimization/yaw_optimizer_sr.py#L15) and finally the base [`YawOptimization` class](https://github.com/NREL/floris/blob/main/floris/optimization/yaw_optimization/yaw_optimization_base.py#L13), which then runs the verification procedure [here](https://github.com/NREL/floris/blob/main/floris/optimization/yaw_optimization/yaw_optimization_base.py#L420) when finalizing the yaw angles. It seems that, for some reason when this is running in parallel, there is a bug in the indexing.

All that being said, we have recently deprecated the `ParallelFlorisModel` class and recommend that you try using its replacement, [`ParFlorisModel`](https://github.com/NREL/floris/blob/main/floris/par_floris_model.py#L13), instead. This class is simply a drop-in replacement for the standard `FlorisModel` that can be passed directly to a yaw optimizer; that is, rather than having a special method `optimize_yaw_angles`, you can pass an instantiated `ParFlorisModel` directly to `YawOptimizationSR`. [See here](https://github.com/NREL/floris/blob/main/examples/examples_control_optimization/005_optimize_yaw_aep_parallel.py) for an example of how to do this.

Because our support has shifted to the more robust and interchangeable `ParFlorisModel`, we may not fix the bug you have identified. However, please do let us know if you run a `YawOptimizationSR` optimization using the `ParFlorisModel` with `verify_convergence=True` and still see the issue!

Best,
Misha

**Comment by fwasilcz** [comment]
Hello Misha, 
Thank you very much for your reply! I was unaware that there is a new class for parallel simulations. I'll try it out, see if problems persist and get back to you. 
Regards
Filip
